File: src/server/orchestrator/branchNameWatcher.ts
import fs from 'node:fs';

const VALID = /^[a-zA-Z0-9]([a-zA-Z0-9_./-]*[a-zA-Z0-9])?$/;
const MAX_LEN = 100;

export interface BranchNameWatcherOptions {
  path: string;
  pollMs?: number;
  onBranchName: (name: string) => void;
  onError: (reason: string) => void;
}

export class BranchNameWatcher {
  private opts: Required<Omit<BranchNameWatcherOptions, 'onBranchName' | 'onError'>> & BranchNameWatcherOptions;
  private timer: NodeJS.Timeout | null = null;
  private running = false;
  private lastEmitted: string | null = null;

  constructor(opts: BranchNameWatcherOptions) {
    this.opts = { pollMs: 1000, ...opts };
  }

  start(): void {
    if (this.running) return;
    this.running = true;
    const tick = () => {
      if (!this.running) return;
      try { this.readOnce(); } catch (e) { this.opts.onError(String(e)); }
      this.timer = setTimeout(tick, this.opts.pollMs);
    };
    this.timer = setTimeout(tick, 0);
  }

  async stop(): Promise<void> {
    this.running = false;
    if (this.timer) { clearTimeout(this.timer); this.timer = null; }
    try { this.readOnce(); } catch (e) { this.opts.onError(String(e)); }
  }

  private readOnce(): void {
    let raw: string;
    try {
      raw = fs.readFileSync(this.opts.path, 'utf8');
    } catch (e) {
      if ((e as NodeJS.ErrnoException).code === 'ENOENT') return;
      throw e;
    }
    const name = raw.trim().slice(0, MAX_LEN);
    if (!name || !VALID.test(name)) return;
    if (name === this.lastEmitted) return;
    this.lastEmitted = name;
    this.opts.onBranchName(name);
  }
}

File: server-elixir/lib/fbi/orchestrator/branch_name_watcher.ex
defmodule FBI.Orchestrator.BranchNameWatcher do
  @moduledoc "Port of src/server/orchestrator/branchNameWatcher.ts."
  use GenServer

  # Mirrors VALID in branchNameWatcher.ts:3 — alphanumeric with optional
  # interior dots, slashes, hyphens, underscores. Not strictly kebab-case;
  # the "2–4 kebab-case words" guidance is preamble-only.
  @valid ~r/^[a-zA-Z0-9]([a-zA-Z0-9_.\/-]*[a-zA-Z0-9])?$/
  @max_len 100

  defstruct [:path, :poll_ms, :on_branch_name, :last_emitted, :timer]

  def start_link(opts) do
    GenServer.start_link(__MODULE__, opts)
  end

  def stop(pid), do: GenServer.stop(pid)

  @impl true
  def init(opts) do
    state = %__MODULE__{
      path: Keyword.fetch!(opts, :path),
      poll_ms: Keyword.get(opts, :poll_ms, 1000),
      on_branch_name: Keyword.fetch!(opts, :on_branch_name),
      last_emitted: nil,
      timer: nil
    }

    {:ok, schedule(state)}
  end

  @impl true
  def handle_info(:tick, state) do
    state = read_once(state)
    {:noreply, schedule(state)}
  end

  @impl true
  def terminate(_reason, state) do
    if state.timer, do: Process.cancel_timer(state.timer)
    read_once(state)
    :ok
  end

  defp schedule(state) do
    timer = Process.send_after(self(), :tick, state.poll_ms)
    %{state | timer: timer}
  end

  defp read_once(%{path: path, on_branch_name: on_branch_name, last_emitted: last} = state) do
    case File.read(path) do
      {:ok, raw} ->
        trimmed = raw |> String.trim() |> String.slice(0, @max_len)

        if trimmed != "" and trimmed != last and Regex.match?(@valid, trimmed) do
          on_branch_name.(trimmed)
          %{state | last_emitted: trimmed}
        else
          state
        end

      {:error, :enoent} ->
        state

      _ ->
        state
    end
  end
end
